### Projeto de Predição da Evolução do Índice IFIX

O Índice IFIX é o resultado de uma carteira teórica de ativos de fundos imobiliários. Seu objetivo é ser o indicador do desempenho médio das cotações dos fundos imobiliários negociados nos mercados de bolsa e de balcão organizado da B3.

> Os dados do índice foram coletados do site da B3 e podem ser acessados [neste link](https://www.b3.com.br/pt_br/market-data-e-indices/indices/indices-de-segmentos-e-setoriais/indice-fundos-de-investimentos-imobiliarios-ifix-estatisticas-historicas.htm).

Para as variáveis preditoras, foram usados dados históricos dos indicadores abaixo: 

* Taxa SELIC: é gerada através da média ponderada das operações diárias titulos públicos, sendo as taxas praticadas balizadas pela SELIC meta.
* CDI: de forma semelhante, é a estabelecido pela média das taxas praticadas entre dos empréstimos diários interbancários. É também balizada pela SELIC meta.
* IGP-M: pesquisado entre os dias 21 do mês anterior e 20 do mês de referência; 
  * Composição: 60% Índice de Preços ao Produtor Amplo (IPA); 30%  Índice de Preços ao Consumidor (IPC); 10%  Índice Nacional de Custo da Construção (INCC). 
  * Abrangência setorial: Alimentação, Habitação, Vestuário, Saúde e Cuidados Pessoais, Educação, Leitura e Recreação, Transportes, Despesas Diversas e Comunicação, sob o ponto de vista do produtor e do consumidor;
Materiais e equipamentos, serviços e mão-de-obra
* INCC: pesquisado entre os dias 21 do mês anterior e 20 do mês de referência; 
  * Abrangência setorial: materiais, serviços e mão-de-obra mais relevantes para a construção civil.

> Dados da SELIC e CDI podem ser obitidos também no site da B3 [neste link](http://estatisticas.cetip.com.br/astec/series_v05/paginas/lum_web_v05_template_informacoes_di.asp?str_Modulo=completo&int_Idioma=1&int_Titulo=6&int_NivelBD=2).

> Dados do IGP-M e INCC podem ser coletados nos lnks a seguir: [IGP-M](https://extra-ibre.fgv.br/IBRE/sitefgvdados/visualizaconsulta.aspx) e [INCC](https://extra-ibre.fgv.br/IBRE/sitefgvdados/visualizaconsulta.aspx). 




In [0]:
from pyspark.sql import SparkSession #importa a biblioteca que cria a seção do spark

In [0]:
#inicia a seção para a utilização do spark
spark = SparkSession.builder.appName("Regressao IFIX").getOrCreate() #cria a seção caso não exista ou obtém a já criada

In [0]:
%fs ls /FileStore/tables

path name size modificationTime dbfs:/FileStore/tables/HisoricoCDI.csv HisoricoCDI.csv 115462 1710016875000 dbfs:/FileStore/tables/HistoricoCDI.csv HistoricoCDI.csv 115462 1710017252000 dbfs:/FileStore/tables/HistoricoIGPM-1.csv HistoricoIGPM-1.csv 8414 1710686210000 dbfs:/FileStore/tables/HistoricoIGPM.csv HistoricoIGPM.csv 6746 1710685001000 dbfs:/FileStore/tables/HistoricoIGP_M.csv HistoricoIGP_M.csv 6746 1710684725000 dbfs:/FileStore/tables/HistoricoINCC-1.csv HistoricoINCC-1.csv 6633 1710686210000 dbfs:/FileStore/tables/HistoricoINCC.csv HistoricoINCC.csv 5305 1710684725000 dbfs:/FileStore/tables/HistoricoSELIC.csv HistoricoSELIC.csv 43851 1710016875000 dbfs:/FileStore/tables/IfixEvolucaoMensal.csv IfixEvolucaoMensal.csv 2347 1710016875000 dbfs:/FileStore/tables/adult_data.csv adult_data.csv 5608318 1709852380000 dbfs:/FileStore/tables/iris_bezdekIris.csv iris_bezdekIris.csv 4551 1709994811000 dbfs:/FileStore/tables/regressaoLinear.csv regressaoLinear.csv 564 1709993226000 dbfs:/FileStore/tables/temperature.csv temperature.csv 4075992 1709677144000 dbfs:/FileStore/tables/temperatureAmericas.csv temperatureAmericas.csv 14644660 1709685210000 dbfs:/FileStore/tables/temperatureMajorCity.csv temperatureMajorCity.csv 14138385 1709678535000 dbfs:/FileStore/tables/u.data u.data 1979173 1710279577000

In [0]:
 # diretórios dos arquivos a serem utilizados
historicoCDI="/FileStore/tables/HistoricoCDI.csv"
historicoSELIC="/FileStore/tables/HistoricoSELIC.csv"
historicoIFIX="/FileStore/tables/IfixEvolucaoMensal.csv"
historicoIGPM="/FileStore/tables/HistoricoIGPM-1.csv"
historicoINCC="/FileStore/tables/HistoricoINCC-1.csv" 

#### Dataset CDI

In [0]:
#lendo arquivos armazenados CSV com o esquema definido
cdi = spark.read.format('csv').options(inferSchema=True, header='true',delimiter=';').load(historicoCDI)

In [0]:
cdi.show(5)

+----------+-------------+-------+-----+------------+
|      Data|Nr. Operações| Volume|Média|Fator Diário|
+----------+-------------+-------+-----+------------+
|2013-01-02|           15|1892473| 6,92|  1,00026555|
|2013-01-03|           16|2007587| 6,92|  1,00026555|
|2013-01-04|           17|2138985| 6,93|  1,00026593|
|2013-01-07|           18|2208986| 6,92|  1,00026555|
|2013-01-08|           18|2246656| 6,91|  1,00026518|
+----------+-------------+-------+-----+------------+
only showing top 5 rows



In [0]:
cdi.printSchema()

root
 |-- Data: date (nullable = true)
 |-- Nr. Operações: integer (nullable = true)
 |-- Volume: string (nullable = true)
 |-- Média: string (nullable = true)
 |-- Fator Diário: string (nullable = true)



In [0]:
import pyspark.sql.functions as F
cdi = cdi.select(F.col('Média').alias("CDI"), F.col('Data').alias("Data"))  # trocando nomes do cabaçalho e selecionando as colunas que preciso

In [0]:
from pyspark.sql.functions import dayofmonth, month, year, last_day
# Adicione uma nova coluna contendo o último dia de cada mês
cdiData = cdi.withColumn("ultimo_dia", last_day("Data"))
# Filtrando pelo último dia de cada mês
cdiData = cdiData.filter(dayofmonth("Data") == dayofmonth("ultimo_dia"))

cdiData.show(5)

+----+----------+----------+
| CDI|      Data|ultimo_dia|
+----+----------+----------+
|6,95|2013-01-31|2013-01-31|
|6,98|2013-02-28|2013-02-28|
|7,23|2013-04-30|2013-04-30|
|7,72|2013-05-31|2013-05-31|
|8,22|2013-07-31|2013-07-31|
+----+----------+----------+
only showing top 5 rows



In [0]:
cdiData = cdiData.select('Data', "CDI")
cdiData.show(5)

+----------+----+
|      Data| CDI|
+----------+----+
|2013-01-31|6,95|
|2013-02-28|6,98|
|2013-04-30|7,23|
|2013-05-31|7,72|
|2013-07-31|8,22|
+----------+----+
only showing top 5 rows



In [0]:
cdiFinal = cdiData.withColumn("CDI", F.regexp_replace(F.col("CDI"), "[,]", "."))  #troca o valor de "," para "." 

In [0]:
cdiFinal.show(5)

+----------+----+
|      Data| CDI|
+----------+----+
|2013-01-31|6.95|
|2013-02-28|6.98|
|2013-04-30|7.23|
|2013-05-31|7.72|
|2013-07-31|8.22|
+----------+----+
only showing top 5 rows



In [0]:
cdiFinal.printSchema()

root
 |-- Data: date (nullable = true)
 |-- CDI: string (nullable = true)



In [0]:
# Trocando CDI para tipo double
from pyspark.sql.functions import col
cdiConcat = cdiFinal.withColumn("CDI", col("CDI").cast("double"))

In [0]:
cdiConcat.printSchema()

root
 |-- Data: date (nullable = true)
 |-- CDI: double (nullable = true)



In [0]:
display(cdiConcat)

Data CDI 2013-01-31 6.95 2013-02-28 6.98 2013-04-30 7.23 2013-05-31 7.72 2013-07-31 8.22 2013-09-30 8.71 2013-10-31 9.32 2013-12-31 9.77 2014-01-31 10.27 2014-02-28 10.59 2014-03-31 10.55 2014-04-30 10.8 2014-06-30 10.8 2014-07-31 10.8 2014-09-30 10.81 2014-10-31 11.06 2014-12-31 11.57 2015-03-31 12.6 2015-04-30 13.11 2015-06-30 13.64 2015-07-31 14.13 2015-08-31 14.13 2015-09-30 14.13 2015-11-30 14.14 2015-12-31 14.14 2016-02-29 14.13 2016-03-31 14.13 2016-05-31 14.13 2016-06-30 14.13 2016-08-31 14.13 2016-09-30 14.13 2016-10-31 13.88 2016-11-30 13.88 2017-01-31 12.88 2017-03-31 12.13 2017-05-31 11.13 2017-06-30 10.14 2017-07-31 9.14 2017-08-31 9.14 2017-10-31 7.39 2017-11-30 7.39 2018-01-31 6.89 2018-02-28 6.64 2018-04-30 6.39 2018-07-31 6.39 2018-08-31 6.39 2018-10-31 6.4 2018-11-30 6.4 2018-12-31 6.4 2019-01-31 6.4 2019-02-28 6.4 2019-04-30 6.4 2019-05-31 6.4 2019-07-31 6.4 2019-09-30 5.4 2019-10-31 4.9 2019-12-31 4.4 2020-01-31 4.4 2020-03-31 3.65 2020-04-30 3.65 2020-06-30 2.15 2020-07-31 2.15 2020-08-31 1.9 2020-09-30 1.9 2020-11-30 1.9 2020-12-31 1.9 2021-03-31 2.65 2021-04-30 2.65 2021-05-31 3.4 2021-06-30 4.15 2021-08-31 5.15 2021-09-30 6.15 2021-11-30 7.65 2021-12-31 9.15 2022-01-31 9.15 2022-03-31 11.65 2022-05-31 12.65 2022-06-30 13.15 2022-08-31 13.65 2022-09-30 13.65 2022-10-31 13.65 2022-11-30 13.65 2023-01-31 13.65 2023-02-28 13.65 2023-03-31 13.65 2023-05-31 13.65 2023-06-30 13.65 2023-07-31 13.65 2023-08-31 13.15 2023-10-31 12.65 2023-11-30 12.15 2024-01-31 11.65 2024-02-29 11.15

Databricks visualization. Run in Databricks to view.

#### Dataset IFIX

In [0]:
# Importanto dataset do IFIX
ifix = spark.read.format('csv').options(inferSchema=True, header='true',delimiter=';').load(historicoIFIX)

In [0]:
ifix.show(5)

+---+----+--------+
|Mês| Ano|   Valor|
+---+----+--------+
|  1|2013|1.613,76|
|  2|2013|1.597,88|
|  3|2013|1.579,32|
|  4|2013|1.549,00|
|  5|2013|1.549,36|
+---+----+--------+
only showing top 5 rows



In [0]:
# Concatenando das colunas Mês e Ano e ajustando ao modelo yyyy-MM-dd
from pyspark.sql.functions import concat, lit, to_date
ifixData = ifix.withColumn("Data", to_date(concat(col("Ano"), lit("-"), col("Mês"), lit("-01"))))


In [0]:
ifixData.show(5)

+---+----+--------+----------+
|Mês| Ano|   Valor|      Data|
+---+----+--------+----------+
|  1|2013|1.613,76|2013-01-01|
|  2|2013|1.597,88|2013-02-01|
|  3|2013|1.579,32|2013-03-01|
|  4|2013|1.549,00|2013-04-01|
|  5|2013|1.549,36|2013-05-01|
+---+----+--------+----------+
only showing top 5 rows



In [0]:
# Alterando a coluna Valor para nome IFIX
ifixData = ifixData.select(F.col('Valor').alias("IFIX"), F.col('Data').alias("Data"))
ifixData.show(5)

+--------+----------+
|    IFIX|      Data|
+--------+----------+
|1.613,76|2013-01-01|
|1.597,88|2013-02-01|
|1.579,32|2013-03-01|
|1.549,00|2013-04-01|
|1.549,36|2013-05-01|
+--------+----------+
only showing top 5 rows



In [0]:
# Retirando o ponto da casa de milhares
from pyspark.sql.functions import regexp_replace
ifixFinal = ifixData.withColumn("IFIX", regexp_replace("IFIX", "\.", ""))
ifixFinal.show(5)

+-------+----------+
|   IFIX|      Data|
+-------+----------+
|1613,76|2013-01-01|
|1597,88|2013-02-01|
|1579,32|2013-03-01|
|1549,00|2013-04-01|
|1549,36|2013-05-01|
+-------+----------+
only showing top 5 rows



In [0]:
# Trocando virgula por ponto
ifixFinal = ifixFinal.withColumn("IFIX", F.regexp_replace(F.col("IFIX"), "[,]", "."))
ifixFinal.show(5)

+-------+----------+
|   IFIX|      Data|
+-------+----------+
|1613.76|2013-01-01|
|1597.88|2013-02-01|
|1579.32|2013-03-01|
|1549.00|2013-04-01|
|1549.36|2013-05-01|
+-------+----------+
only showing top 5 rows



In [0]:
ifixFinal.printSchema()

root
 |-- IFIX: string (nullable = true)
 |-- Data: date (nullable = true)



In [0]:
ifixConcat = ifixFinal.withColumn("IFIX", col("IFIX").cast("Double")) # trocando valor string para Double

In [0]:
ifixConcat.printSchema()

root
 |-- IFIX: double (nullable = true)
 |-- Data: date (nullable = true)



In [0]:
ifixConcat.show(5)

+-------+----------+
|   IFIX|      Data|
+-------+----------+
|1613.76|2013-01-01|
|1597.88|2013-02-01|
|1579.32|2013-03-01|
| 1549.0|2013-04-01|
|1549.36|2013-05-01|
+-------+----------+
only showing top 5 rows



In [0]:
display(ifixConcat)

IFIX Data 1613.76 2013-01-01 1597.88 2013-02-01 1579.32 2013-03-01 1549.0 2013-04-01 1549.36 2013-05-01 1437.76 2013-06-01 1427.11 2013-07-01 1375.07 2013-08-01 1400.93 2013-09-01 1423.1 2013-10-01 1411.67 2013-11-01 1374.7 2013-12-01 1282.04 2014-01-01 1330.42 2014-02-01 1336.07 2014-03-01 1350.76 2014-04-01 1369.25 2014-05-01 1387.63 2014-06-01 1409.42 2014-07-01 1413.1 2014-08-01 1421.44 2014-09-01 1402.46 2014-10-01 1361.51 2014-11-01 1336.72 2014-12-01 1372.06 2015-01-01 1369.42 2015-02-01 1347.63 2015-03-01 1397.06 2015-04-01 1417.54 2015-05-01 1460.46 2015-06-01 1470.73 2015-07-01 1458.12 2015-08-01 1400.42 2015-09-01 1430.17 2015-10-01 1452.66 2015-11-01 1409.01 2015-12-01 1322.95 2016-01-01 1361.05 2016-02-01 1485.75 2016-03-01 1554.17 2016-04-01 1612.36 2016-05-01 1638.16 2016-06-01 1735.99 2016-07-01 1767.11 2016-08-01 1816.05 2016-09-01 1885.92 2016-10-01 1837.13 2016-11-01 1864.61 2016-12-01 1934.69 2017-01-01 2028.86 2017-02-01 2032.81 2017-03-01 2035.85 2017-04-01 2056.75 2017-05-01 2074.14 2017-06-01 2066.19 2017-07-01 2084.0 2017-08-01 2221.05 2017-09-01 2226.25 2017-10-01 2213.08 2017-11-01 2226.46 2017-12-01 2285.23 2018-01-01 2311.41 2018-02-01 2357.54 2018-03-01 2337.25 2018-04-01 2214.11 2018-05-01 2125.22 2018-06-01 2154.45 2018-07-01 2139.4 2018-08-01 2134.81 2018-09-01 2242.45 2018-10-01 2300.48 2018-11-01 2351.59 2018-12-01 2409.69 2019-01-01 2434.56 2019-02-01 2482.93 2019-03-01 2508.45 2019-04-01 2552.69 2019-05-01 2626.1 2019-06-01 2659.55 2019-07-01 2656.69 2019-08-01 2684.26 2019-09-01 2791.98 2019-10-01 2890.26 2019-11-01 3197.58 2019-12-01 3077.28 2020-01-01 2963.81 2020-02-01 2494.1 2020-03-01 2603.62 2020-04-01 2657.68 2020-05-01 2806.28 2020-06-01 2733.12 2020-07-01 2782.1 2020-08-01 2794.88 2020-09-01 2766.76 2020-10-01 2808.56 2020-11-01 2870.15 2020-12-01 2879.45 2021-01-01 2886.53 2021-02-01 2846.77 2021-03-01 2861.15 2021-04-01 2816.46 2021-05-01 2754.89 2021-06-01 2824.12 2021-07-01 2749.84 2021-08-01 2715.67 2021-09-01 2675.68 2021-10-01 2578.4 2021-11-01 2804.79 2021-12-01 2776.92 2022-01-01 2741.15 2022-02-01 2779.97 2022-03-01 2813.06 2022-04-01 2820.45 2022-05-01 2795.62 2022-06-01 2814.1 2022-07-01 2976.23 2022-08-01 2990.8 2022-09-01 2991.45 2022-10-01 2867.21 2022-11-01 2867.13 2022-12-01 2821.14 2023-01-01 2808.39 2023-02-01 2760.96 2023-03-01 2858.23 2023-04-01 3013.46 2023-05-01 3155.32 2023-06-01 3197.18 2023-07-01 3212.81 2023-08-01 3219.28 2023-09-01 3155.7 2023-10-01 3176.56 2023-11-01 3311.43 2023-12-01 3333.51 2024-01-01 3360.0 2024-02-01 3382.28 2024-03-01

Databricks visualization. Run in Databricks to view.

In [0]:
ifixConcat.describe().show()

+-------+-----------------+
|summary|             IFIX|
+-------+-----------------+
|  count|              135|
|   mean|2243.024074074075|
| stddev|647.2107060950273|
|    min|          1282.04|
|    max|          3382.28|
+-------+-----------------+



#### Dataset SELIC  

In [0]:
selic = spark.read.format('csv').options(inferSchema=True, header='true',delimiter=';').load(historicoSELIC)

In [0]:
selic.show(5)

+----------+----------+
|      Data|Taxa SELIC|
+----------+----------+
|2014-01-02|       9,9|
|2014-01-03|       9,9|
|2014-01-06|       9,9|
|2014-01-07|       9,9|
|2014-01-08|       9,9|
+----------+----------+
only showing top 5 rows



In [0]:
selic.printSchema()

root
 |-- Data: date (nullable = true)
 |-- Taxa SELIC: string (nullable = true)



In [0]:
# Renomeando a coluna Taxa SELIC para apenas SELIC
selic = selic.select(F.col('Taxa SELIC').alias("SELIC"), F.col('Data').alias("Data"))
selic.show(5)

+-----+----------+
|SELIC|      Data|
+-----+----------+
|  9,9|2014-01-02|
|  9,9|2014-01-03|
|  9,9|2014-01-06|
|  9,9|2014-01-07|
|  9,9|2014-01-08|
+-----+----------+
only showing top 5 rows



In [0]:
# Trocando virgula por ponto
selic = selic.withColumn("SELIC", F.regexp_replace(F.col("SELIC"), "[,]", ".")) # trocando ',' por '.'
selic.show(5)

+-----+----------+
|SELIC|      Data|
+-----+----------+
|  9.9|2014-01-02|
|  9.9|2014-01-03|
|  9.9|2014-01-06|
|  9.9|2014-01-07|
|  9.9|2014-01-08|
+-----+----------+
only showing top 5 rows



In [0]:
# Adicione uma nova coluna contendo o último dia de cada mês
selicData = selic.withColumn("ultimo_dia", last_day("Data"))
# Filtrando pelo último dia de cada mês
selicData = selicData.filter(dayofmonth("Data") == dayofmonth("ultimo_dia"))
selicData.show(5)

+-----+----------+----------+
|SELIC|      Data|ultimo_dia|
+-----+----------+----------+
| 10.4|2014-01-31|2014-01-31|
|10.65|2014-02-28|2014-02-28|
|10.65|2014-03-31|2014-03-31|
| 10.9|2014-04-30|2014-04-30|
| 10.9|2014-06-30|2014-06-30|
+-----+----------+----------+
only showing top 5 rows



In [0]:
# Selecionando as colunas neessárias
selicFinal = selicData.select('Data', 'SELIC')
selicFinal.printSchema()

root
 |-- Data: date (nullable = true)
 |-- SELIC: string (nullable = true)



In [0]:
# trocando valor string para double
selicConcat = selicFinal.withColumn("SELIC", col("SELIC").cast("double")) 
display(selicConcat)

Data SELIC 2014-01-31 10.4 2014-02-28 10.65 2014-03-31 10.65 2014-04-30 10.9 2014-06-30 10.9 2014-07-31 10.9 2014-09-30 10.9 2014-10-31 11.15 2014-12-31 11.65 2015-03-31 12.65 2015-04-30 13.15 2015-06-30 13.65 2015-07-31 14.15 2015-08-31 14.15 2015-09-30 14.15 2015-11-30 14.15 2015-12-31 14.15 2016-02-29 14.15 2016-03-31 14.15 2016-05-31 14.15 2016-06-30 14.15 2016-08-31 14.15 2016-09-30 14.15 2016-10-31 13.9 2016-11-30 13.9 2017-01-31 12.9 2017-03-31 12.15 2017-05-31 11.15 2017-06-30 10.15 2017-07-31 9.15 2017-08-31 9.15 2017-10-31 7.4 2017-11-30 7.4 2018-01-31 6.9 2018-02-28 6.65 2018-04-30 6.4 2018-07-31 6.4 2018-08-31 6.4 2018-10-31 6.4 2018-11-30 6.4 2018-12-31 6.4 2019-01-31 6.4 2019-02-28 6.4 2019-04-30 6.4 2019-05-31 6.4 2019-07-31 6.4 2019-09-30 5.4 2019-10-31 4.9 2019-12-31 4.4 2020-01-31 4.4 2020-03-31 3.65 2020-04-30 3.65 2020-06-30 2.15 2020-07-31 2.15 2020-08-31 1.9 2020-09-30 1.9 2020-11-30 1.9 2020-12-31 1.9 2021-03-31 2.65 2021-04-30 2.65 2021-05-31 3.4 2021-06-30 4.15 2021-08-31 5.15 2021-09-30 6.15 2021-11-30 7.65 2021-12-31 9.15 2022-01-31 9.15 2022-03-31 11.65 2022-05-31 12.65 2022-06-30 13.15 2022-08-31 13.65 2022-09-30 13.65 2022-10-31 13.65 2022-11-30 13.65 2023-01-31 13.65 2023-02-28 13.65 2023-03-31 13.65 2023-05-31 13.65 2023-06-30 13.65 2023-07-31 13.65 2023-08-31 13.15 2023-10-31 12.65 2023-11-30 12.15 2024-01-31 11.65 2024-02-29 11.15

Databricks visualization. Run in Databricks to view.

In [0]:
selicConcat.printSchema()

root
 |-- Data: date (nullable = true)
 |-- SELIC: double (nullable = true)



#### Dataset IGP-M

In [0]:
igpm = spark.read.format('csv').options(inferSchema=True, header='true',delimiter=';').load(historicoIGPM)

In [0]:
display(igpm)

Data IGP-M 01/06/1989 3,09E-05 01/07/1989 4,19E-05 01/08/1989 5,74E-05 01/09/1989 8,03E-05 01/10/1989 1,13E-04 01/11/1989 1,59E-04 01/12/1989 2,34E-04 01/01/1990 3,77E-04 01/02/1990 6,84E-04 01/03/1990 1,26E-03 01/04/1990 1,61E-03 01/05/1990 1,71E-03 01/06/1990 1,88E-03 01/07/1990 2,11E-03 01/08/1990 2,39E-03 01/09/1990 2,70E-03 01/10/1990 3,05E-03 01/11/1990 3,56E-03 01/12/1990 4,20E-03 01/01/1991 4,95E-03 01/02/1991 5,99E-03 01/03/1991 6,54E-03 01/04/1991 7,05E-03 01/05/1991 7,58E-03 01/06/1991 8,22E-03 01/07/1991 9,30E-03 01/08/1991 1,07E-02 01/09/1991 1,23E-02 01/10/1991 1,51E-02 01/11/1991 1,90E-02 01/12/1991 2,35E-02 01/01/1992 2,90E-02 01/02/1992 3,71E-02 01/03/1992 4,50E-02 01/04/1992 5,40E-02 01/05/1992 6,50E-02 01/06/1992 8,04E-02 01/07/1992 9,79E-02 01/08/1992 1,22E-01 01/09/1992 1,53E-01 01/10/1992 1,94E-01 01/11/1992 2,39E-01 01/12/1992 2,99E-01 01/01/1993 3,76E-01 01/02/1993 4,83E-01 01/03/1993 6,10E-01 01/04/1993 7,86E-01 01/05/1993 1,02E+00 01/06/1993 1,34E+00 01/07/1993 1,76E+00 01/08/1993 2,32E+00 01/09/1993 3,14E+00 01/10/1993 4,24E+00 01/11/1993 5,77E+00 01/12/1993 7,98E+00 01/01/1994 1,11E+01 01/02/1994 1,56E+01 01/03/1994 2,28E+01 01/04/1994 3,21E+01 01/05/1994 4,57E+01 01/06/1994 6,64E+01 01/07/1994 9,30E+01 01/08/1994 100 01/09/1994 101,751 01/10/1994 103,602 01/11/1994 106,553 01/12/1994 107,45 01/01/1995 108,442 01/02/1995 109,945 01/03/1995 111,178 01/04/1995 113,518 01/05/1995 114,171 01/06/1995 116,984 01/07/1995 119,114 01/08/1995 121,729 01/09/1995 120,869 01/10/1995 121,503 01/11/1995 122,955 01/12/1995 123,833 01/01/1996 125,977 01/02/1996 127,202 01/03/1996 127,715 01/04/1996 128,13 01/05/1996 130,121 01/06/1996 131,445 01/07/1996 133,213 01/08/1996 133,587 01/09/1996 133,722 01/10/1996 133,978 01/11/1996 134,242 01/12/1996 135,225 01/01/1997 137,613 01/02/1997 138,204 01/03/1997 139,795 01/04/1997 140,742 01/05/1997 141,04 01/06/1997 142,09 01/07/1997 142,221 01/08/1997 142,353 01/09/1997 143,042 01/10/1997 143,567 01/11/1997 144,481 01/12/1997 145,695 01/01/1998 147,091 01/02/1998 147,356 01/03/1998 147,635 01/04/1998 147,821 01/05/1998 148,021 01/06/1998 148,588 01/07/1998 148,339 01/08/1998 148,109 01/09/1998 147,984 01/10/1998 148,1 01/11/1998 147,628 01/12/1998 148,291 01/01/1999 149,533 01/02/1999 154,933 01/03/1999 159,325 01/04/1999 160,459 01/05/1999 159,996 01/06/1999 160,573 01/07/1999 163,06 01/08/1999 165,603 01/09/1999 167,997 01/10/1999 170,861 01/11/1999 174,939 01/12/1999 178,099 01/01/2000 180,301 01/02/2000 180,935 01/03/2000 181,214 01/04/2000 181,635 01/05/2000 182,189 01/06/2000 183,745 01/07/2000 186,634 01/08/2000 191,087 01/09/2000 193,297 01/10/2000 194,04 01/11/2000 194,599 01/12/2000 195,827 01/01/2001 197,045 01/02/2001 197,491 01/03/2001 198,606 01/04/2001 200,591 01/05/2001 202,324 01/06/2001 204,31 01/07/2001 207,341 01/08/2001 210,211 01/09/2001 210,853 01/10/2001 213,339 01/11/2001 215,685 01/12/2001 216,163 01/01/2002 216,944 01/02/2002 217,074 01/03/2002 217,276 01/04/2002 218,486 01/05/2002 220,292 01/06/2002 223,688 01/07/2002 228,057 01/08/2002 233,348 01/09/2002 238,943 01/10/2002 248,199 01/11/2002 261,08 01/12/2002 270,867 01/01/2003 277,173 01/02/2003 283,506 01/03/2003 287,855 01/04/2003 290,512 01/05/2003 289,747 01/06/2003 286,843 01/07/2003 285,649 01/08/2003 286,735 01/09/2003 290,127 01/10/2003 291,229 01/11/2003 292,657 01/12/2003 294,455 01/01/2004 297,039 01/02/2004 299,097 01/03/2004 302,484 01/04/2004 306,151 01/05/2004 310,152 01/06/2004 314,419 01/07/2004 318,532 01/08/2004 322,412 01/09/2004 324,651 01/10/2004 325,925 01/11/2004 328,588 01/12/2004 331,005 01/01/2005 332,298 01/02/2005 333,288 01/03/2005 336,123 01/04/2005 339,03 01/05/2005 338,299 01/06/2005 336,801 01/07/2005 335,663 01/08/2005 333,474 01/09/2005 331,69 01/10/2005 333,694 01/11/2005 335,033 01/12/2005 335,006 01/01/2006 338,083 01/02/2006 338,128 01/03/2006 337,339 01/04/2006 335,921 01/05/2006 337,185 01/06/2006 339,712 01/07/2006 340,312 01/08/2006 341,574 01/09/2006

In [0]:
igpm.printSchema()

root
 |-- Data: string (nullable = true)
 |-- IGP-M: string (nullable = true)



In [0]:
from pyspark.sql.functions import col, to_date, concat_ws, lit, date_format
from pyspark.sql.types import DateType

igpmData = igpm.withColumn("Data", to_date(col("Data"), "dd/MM/yyyy"))

igpmData.show(5)

+----------+--------+
|      Data|   IGP-M|
+----------+--------+
|1989-06-01|3,09E-05|
|1989-07-01|4,19E-05|
|1989-08-01|5,74E-05|
|1989-09-01|8,03E-05|
|1989-10-01|1,13E-04|
+----------+--------+
only showing top 5 rows



In [0]:
# Filtrando valores a partir do ano de 2013
igpmData = igpmData.filter(year(col("Data")) >= 2013)
igpmData.show(5)

+----------+-------+
|      Data|  IGP-M|
+----------+-------+
|2013-01-01|511,977|
|2013-02-01|513,467|
|2013-03-01|514,526|
|2013-04-01|515,276|
|2013-05-01|515,299|
+----------+-------+
only showing top 5 rows



In [0]:
igpmData.printSchema()

root
 |-- Data: date (nullable = true)
 |-- IGP-M: string (nullable = true)



In [0]:
from pyspark.sql.functions import regexp_replace
igpmValor = igpmData.withColumn("IGP-M", F.regexp_replace(F.col("IGP-M"), "[,]", "."))
igpmValor.show(5)
                                

+----------+-------+
|      Data|  IGP-M|
+----------+-------+
|2013-01-01|511.977|
|2013-02-01|513.467|
|2013-03-01|514.526|
|2013-04-01|515.276|
|2013-05-01|515.299|
+----------+-------+
only showing top 5 rows



In [0]:
igpmValor.printSchema()

root
 |-- Data: date (nullable = true)
 |-- IGP-M: string (nullable = true)



In [0]:
igpmFinal = igpmValor.withColumn("IGP-M", col("IGP-M").cast("float"))
igpmFinal.show(5)

+----------+-------+
|      Data|  IGP-M|
+----------+-------+
|2013-01-01|511.977|
|2013-02-01|513.467|
|2013-03-01|514.526|
|2013-04-01|515.276|
|2013-05-01|515.299|
+----------+-------+
only showing top 5 rows



In [0]:
display(igpmFinal)

Data IGP-M 2013-01-01 511.977 2013-02-01 513.467 2013-03-01 514.526 2013-04-01 515.276 2013-05-01 515.299 2013-06-01 519.153 2013-07-01 520.508 2013-08-01 521.27 2013-09-01 529.085 2013-10-01 533.621 2013-11-01 535.168 2013-12-01 538.37 2014-01-01 540.959 2014-02-01 543.038 2014-03-01 552.087 2014-04-01 556.42 2014-05-01 555.679 2014-06-01 551.554 2014-07-01 548.202 2014-08-01 546.745 2014-09-01 547.839 2014-10-01 549.396 2014-11-01 554.769 2014-12-01 558.213 2015-01-01 562.482 2015-02-01 564.004 2015-03-01 569.536 2015-04-01 576.175 2015-05-01 578.516 2015-06-01 582.401 2015-07-01 586.426 2015-08-01 588.042 2015-09-01 593.606 2015-10-01 604.832 2015-11-01 614.051 2015-12-01 617.044 2016-01-01 624.06 2016-02-01 632.114 2016-03-01 635.349 2016-04-01 637.434 2016-05-01 642.651 2016-06-01 653.496 2016-07-01 654.641 2016-08-01 655.602 2016-09-01 656.894 2016-10-01 657.927 2016-11-01 657.752 2016-12-01 661.304 2017-01-01 665.542 2017-02-01 666.099 2017-03-01 666.197 2017-04-01 658.898 2017-05-01 652.758 2017-06-01 648.409 2017-07-01 643.766 2017-08-01 644.383 2017-09-01 647.4 2017-10-01 648.672 2017-11-01 652.073 2017-12-01 657.859 2018-01-01 662.826 2018-02-01 663.311 2018-03-01 667.524 2018-04-01 671.327 2018-05-01 680.579 2018-06-01 693.287 2018-07-01 696.8 2018-08-01 701.677 2018-09-01 712.373 2018-10-01 718.684 2018-11-01 715.166 2018-12-01 707.441 2019-01-01 707.488 2019-02-01 713.747 2019-03-01 722.707 2019-04-01 729.346 2019-05-01 732.595 2019-06-01 738.421 2019-07-01 741.346 2019-08-01 736.402 2019-09-01 736.362 2019-10-01 741.333 2019-11-01 743.558 2019-12-01 759.112 2020-01-01 762.733 2020-02-01 762.423 2020-03-01 771.908 2020-04-01 778.101 2020-05-01 780.28 2020-06-01 792.429 2020-07-01 810.083 2020-08-01 832.313 2020-09-01 868.442 2020-10-01 896.505 2020-11-01 925.887 2020-12-01 934.758 2021-01-01 958.844 2021-02-01 983.063 2021-03-01 1011.948 2021-04-01 1027.211 2021-05-01 1069.289 2021-06-01 1075.733 2021-07-01 1084.095 2021-08-01 1091.29 2021-09-01 1084.312 2021-10-01 1091.283 2021-11-01 1091.483 2021-12-01 1100.988 2022-01-01 1120.999 2022-02-01 1141.546 2022-03-01 1161.418 2022-04-01 1177.809 2022-05-01 1183.953 2022-06-01 1190.882 2022-07-01 1193.337 2022-08-01 1185.004 2022-09-01 1173.793 2022-10-01 1162.391 2022-11-01 1155.829 2022-12-01 1161.006 2023-01-01 1163.465 2023-02-01 1162.761 2023-03-01 1163.359 2023-04-01 1152.307 2023-05-01 1131.058 2023-06-01 1109.23 2023-07-01 1101.204 2023-08-01 1099.71 2023-09-01 1103.74 2023-10-01 1109.236 2023-11-01 1115.815 2023-12-01 1124.072 2024-01-01 1124.879 2024-02-01 1119.061

Databricks visualization. Run in Databricks to view.

In [0]:
igpmFinal.describe().show()

+-------+------------------+
|summary|             IGP-M|
+-------+------------------+
|  count|               134|
|   mean| 783.5648721723414|
| stddev|228.13797633515642|
|    min|           511.977|
|    max|          1193.337|
+-------+------------------+



#### Dataset INCC

In [0]:
incc = spark.read.format('csv').options(inferSchema=True, header='true', delimiter=';').load(historicoINCC)

In [0]:
display(incc)

Data INCC 01/07/1996 149,224 01/08/1996 150,334 01/09/1996 150,564 01/10/1996 150,915 01/11/1996 151,345 01/12/1996 152,079 01/01/1997 153,107 01/02/1997 153,588 01/03/1997 154,228 01/04/1997 155,358 01/05/1997 155,823 01/06/1997 157,373 01/07/1997 158,855 01/08/1997 159,724 01/09/1997 161,604 01/10/1997 161,953 01/11/1997 162,317 01/12/1997 163,165 01/01/1998 163,644 01/02/1998 164,428 01/03/1998 164,935 01/04/1998 164,169 01/05/1998 164,933 01/06/1998 166,356 01/07/1998 167,063 01/08/1998 167,554 01/09/1998 167,698 01/10/1998 167,78 01/11/1998 167,662 01/12/1998 167,652 01/01/1999 168,008 01/02/1999 169,054 01/03/1999 170,587 01/04/1999 171,584 01/05/1999 172,233 01/06/1999 173,75 01/07/1999 174,466 01/08/1999 175,37 01/09/1999 176,824 01/10/1999 178,217 01/11/1999 180,382 01/12/1999 181,839 01/01/2000 183,929 01/02/2000 185,572 01/03/2000 187,092 01/04/2000 188,389 01/05/2000 189,463 01/06/2000 191,521 01/07/2000 193,104 01/08/2000 193,786 01/09/2000 194,365 01/10/2000 194,914 01/11/2000 195,488 01/12/2000 196,45 01/01/2001 197,233 01/02/2001 198,325 01/03/2001 198,996 01/04/2001 199,545 01/05/2001 203,53 01/06/2001 204,931 01/07/2001 207,117 01/08/2001 208,272 01/09/2001 209,51 01/10/2001 211,431 01/11/2001 212,7 01/12/2001 214,194 01/01/2002 215,058 01/02/2002 215,734 01/03/2002 217,554 01/04/2002 218,259 01/05/2002 223,644 01/06/2002 224,117 01/07/2002 225,529 01/08/2002 227,376 01/09/2002 228,913 01/10/2002 230,787 01/11/2002 235,843 01/12/2002 240,868 01/01/2003 244,354 01/02/2003 248,262 01/03/2003 251,695 01/04/2003 253,744 01/05/2003 261,311 01/06/2003 263,244 01/07/2003 264,785 01/08/2003 270,623 01/09/2003 271,284 01/10/2003 272,552 01/11/2003 273,706 01/12/2003 276,424 01/01/2004 277,201 01/02/2004 278,521 01/03/2004 282,956 01/04/2004 284,665 01/05/2004 289,627 01/06/2004 291,239 01/07/2004 294,509 01/08/2004 297,162 01/09/2004 299,14 01/10/2004 301,975 01/11/2004 304,81 01/12/2004 306,655 01/01/2005 308,802 01/02/2005 310,105 01/03/2005 312,3 01/04/2005 313,497 01/05/2005 315,2 01/06/2005 322,15 01/07/2005 324,236 01/08/2005 324,394 01/09/2005 324,575 01/10/2005 325,485 01/11/2005 326,416 01/12/2005 327,645 01/01/2006 328,427 01/02/2006 329,36 01/03/2006 330,108 01/04/2006 330,787 01/05/2006 333,459 01/06/2006 338,302 01/07/2006 340,232 01/08/2006 341,438 01/09/2006 341,731 01/10/2006 342,352 01/11/2006 343,137 01/12/2006 344,171 01/01/2007 345,713 01/02/2007 346,629 01/03/2007 347,227 01/04/2007 348,729 01/05/2007 350,631 01/06/2007 356,503 01/07/2007 357,269 01/08/2007 358,534 01/09/2007 359,916 01/10/2007 361,672 01/11/2007 363,397 01/12/2007 364,972 01/01/2008 366,471 01/02/2008 368,062 01/03/2008 370,251 01/04/2008 373,302 01/05/2008 377,419 01/06/2008 387,512 01/07/2008 393,018 01/08/2008 398,015 01/09/2008 401,797 01/10/2008 405,212 01/11/2008 407,86 01/12/2008 408,765 01/01/2009 409,811 01/02/2009 411,249 01/03/2009 410,569 01/04/2009 410,531 01/05/2009 411,57 01/06/2009 417,86 01/07/2009 419,422 01/08/2009 419,468 01/09/2009 419,758 01/10/2009 420,298 01/11/2009 421,07 01/12/2009 421,908 01/01/2010 424,094 01/02/2010 425,59 01/03/2010 427,498 01/04/2010 432,491 01/05/2010 436,499 01/06/2010 444,243 01/07/2010 446,992 01/08/2010 447,996 01/09/2010 448,892 01/10/2010 449,587 01/11/2010 451,215 01/12/2010 453,876 01/01/2011 455,562 01/02/2011 457,333 01/03/2011 459,36 01/04/2011 462,793 01/05/2011 472,203 01/06/2011 478,935 01/07/2011 481,768 01/08/2011 482,518 01/09/2011 483,199 01/10/2011 484,185 01/11/2011 486,61 01/12/2011 488,298 01/01/2012 491,549 01/02/2012 493,608 01/03/2012 495,433 01/04/2012 499,548 01/05/2012 506,02 01/06/2012 512,628 01/07/2012 516,963 01/08/2012 518,616 01/09/2012 519,681 01/10/2012 520,94 01/11/2012 522,136 01/12/2012 523,624 01/01/2013 525,651 01/02/2013 529,879 01/03/2013 531,339 01/04/2013 535,777 01/05/2013 542,402 01/06/2013 553,007 01/07/2013 557,04 01/08/2013 558,777 01/09/2013 561,203 01/10/2013 563,034 01/11/2013 564,558 01/12/2013 565,786 01/01/2014 569,745 01/02/2

In [0]:
incc.printSchema()

root
 |-- Data: string (nullable = true)
 |-- INCC: string (nullable = true)



In [0]:
inccData = incc.withColumn("Data", to_date(col("Data"), "dd/MM/yyyy"))
inccData.show(5)

+----------+-------+
|      Data|   INCC|
+----------+-------+
|1996-07-01|149,224|
|1996-08-01|150,334|
|1996-09-01|150,564|
|1996-10-01|150,915|
|1996-11-01|151,345|
+----------+-------+
only showing top 5 rows



In [0]:
inccValor = inccData.withColumn("INCC", F.regexp_replace(F.col("INCC"), "[,]", "."))
inccValor.show(5)

+----------+-------+
|      Data|   INCC|
+----------+-------+
|1996-07-01|149.224|
|1996-08-01|150.334|
|1996-09-01|150.564|
|1996-10-01|150.915|
|1996-11-01|151.345|
+----------+-------+
only showing top 5 rows



In [0]:
inccFinal = inccValor.withColumn("INCC", col("INCC").cast("float"))
inccFinal.show(5)

+----------+-------+
|      Data|   INCC|
+----------+-------+
|1996-07-01|149.224|
|1996-08-01|150.334|
|1996-09-01|150.564|
|1996-10-01|150.915|
|1996-11-01|151.345|
+----------+-------+
only showing top 5 rows



In [0]:
inccFinal.describe().show()

+-------+------------------+
|summary|              INCC|
+-------+------------------+
|  count|               332|
|   mean| 492.1975475908762|
| stddev|269.02511345000295|
|    min|           149.224|
|    max|          1090.871|
+-------+------------------+



In [0]:
display(inccFinal)

Data INCC 1996-07-01 149.224 1996-08-01 150.334 1996-09-01 150.564 1996-10-01 150.915 1996-11-01 151.345 1996-12-01 152.079 1997-01-01 153.107 1997-02-01 153.588 1997-03-01 154.228 1997-04-01 155.358 1997-05-01 155.823 1997-06-01 157.373 1997-07-01 158.855 1997-08-01 159.724 1997-09-01 161.604 1997-10-01 161.953 1997-11-01 162.317 1997-12-01 163.165 1998-01-01 163.644 1998-02-01 164.428 1998-03-01 164.935 1998-04-01 164.169 1998-05-01 164.933 1998-06-01 166.356 1998-07-01 167.063 1998-08-01 167.554 1998-09-01 167.698 1998-10-01 167.78 1998-11-01 167.662 1998-12-01 167.652 1999-01-01 168.008 1999-02-01 169.054 1999-03-01 170.587 1999-04-01 171.584 1999-05-01 172.233 1999-06-01 173.75 1999-07-01 174.466 1999-08-01 175.37 1999-09-01 176.824 1999-10-01 178.217 1999-11-01 180.382 1999-12-01 181.839 2000-01-01 183.929 2000-02-01 185.572 2000-03-01 187.092 2000-04-01 188.389 2000-05-01 189.463 2000-06-01 191.521 2000-07-01 193.104 2000-08-01 193.786 2000-09-01 194.365 2000-10-01 194.914 2000-11-01 195.488 2000-12-01 196.45 2001-01-01 197.233 2001-02-01 198.325 2001-03-01 198.996 2001-04-01 199.545 2001-05-01 203.53 2001-06-01 204.931 2001-07-01 207.117 2001-08-01 208.272 2001-09-01 209.51 2001-10-01 211.431 2001-11-01 212.7 2001-12-01 214.194 2002-01-01 215.058 2002-02-01 215.734 2002-03-01 217.554 2002-04-01 218.259 2002-05-01 223.644 2002-06-01 224.117 2002-07-01 225.529 2002-08-01 227.376 2002-09-01 228.913 2002-10-01 230.787 2002-11-01 235.843 2002-12-01 240.868 2003-01-01 244.354 2003-02-01 248.262 2003-03-01 251.695 2003-04-01 253.744 2003-05-01 261.311 2003-06-01 263.244 2003-07-01 264.785 2003-08-01 270.623 2003-09-01 271.284 2003-10-01 272.552 2003-11-01 273.706 2003-12-01 276.424 2004-01-01 277.201 2004-02-01 278.521 2004-03-01 282.956 2004-04-01 284.665 2004-05-01 289.627 2004-06-01 291.239 2004-07-01 294.509 2004-08-01 297.162 2004-09-01 299.14 2004-10-01 301.975 2004-11-01 304.81 2004-12-01 306.655 2005-01-01 308.802 2005-02-01 310.105 2005-03-01 312.3 2005-04-01 313.497 2005-05-01 315.2 2005-06-01 322.15 2005-07-01 324.236 2005-08-01 324.394 2005-09-01 324.575 2005-10-01 325.485 2005-11-01 326.416 2005-12-01 327.645 2006-01-01 328.427 2006-02-01 329.36 2006-03-01 330.108 2006-04-01 330.787 2006-05-01 333.459 2006-06-01 338.302 2006-07-01 340.232 2006-08-01 341.438 2006-09-01 341.731 2006-10-01 342.352 2006-11-01 343.137 2006-12-01 344.171 2007-01-01 345.713 2007-02-01 346.629 2007-03-01 347.227 2007-04-01 348.729 2007-05-01 350.631 2007-06-01 356.503 2007-07-01 357.269 2007-08-01 358.534 2007-09-01 359.916 2007-10-01 361.672 2007-11-01 363.397 2007-12-01 364.972 2008-01-01 366.471 2008-02-01 368.062 2008-03-01 370.251 2008-04-01 373.302 2008-05-01 377.419 2008-06-01 387.512 2008-07-01 393.018 2008-08-01 398.015 2008-09-01 401.797 2008-10-01 405.212 2008-11-01 407.86 2008-12-01 408.765 2009-01-01 409.811 2009-02-01 411.249 2009-03-01 410.569 2009-04-01 410.531 2009-05-01 411.57 2009-06-01 417.86 2009-07-01 419.422 2009-08-01 419.468 2009-09-01 419.758 2009-10-01 420.298 2009-11-01 421.07 2009-12-01 421.908 2010-01-01 424.094 2010-02-01 425.59 2010-03-01 427.498 2010-04-01 432.491 2010-05-01 436.499 2010-06-01 444.243 2010-07-01 446.992 2010-08-01 447.996 2010-09-01 448.892 2010-10-01 449.587 2010-11-01 451.215 2010-12-01 453.876 2011-01-01 455.562 2011-02-01 457.333 2011-03-01 459.36 2011-04-01 462.793 2011-05-01 472.203 2011-06-01 478.935 2011-07-01 481.768 2011-08-01 482.518 2011-09-01 483.199 2011-10-01 484.185 2011-11-01 486.61 2011-12-01 488.298 2012-01-01 491.549 2012-02-01 493.608 2012-03-01 495.433 2012-04-01 499.548 2012-05-01 506.02 2012-06-01 512.628 2012-07-01 516.963 2012-08-01 518.616 2012-09-01 519.681 2012-10-01 520.94 2012-11-01 522.136 2012-12-01 523.624 2013-01-01 525.651 2013-02-01 529.879 2013-03-01 531.339 2013-04-01 535.777 2013-05-01 542.402 2013-06-01 553.007 2013-07-01 557.04 2013-08-01 558.777 2013-09-01 561.203 2013-10-01 563.034 2013-11-01 564.558 2013-12-01 565.786 2014-01-01 569.745 2014-02

Databricks visualization. Run in Databricks to view.

# Concatenando Datasets

#### Ajustando os datasets

In [0]:
# Extraindo mês e ano da coluna de data
ifixReg = ifixConcat.withColumn("Mes_Ano", concat(month("Data"), lit("-"), year("Data")))
ifixReg = ifixReg.select('IFIX', 'Mes_Ano')
ifixReg.show(5)

+-------+-------+
|   IFIX|Mes_Ano|
+-------+-------+
|1613.76| 1-2013|
|1597.88| 2-2013|
|1579.32| 3-2013|
| 1549.0| 4-2013|
|1549.36| 5-2013|
+-------+-------+
only showing top 5 rows



In [0]:
# Extraindo mês e ano da coluna de data
cdiReg = cdiConcat.withColumn("Mes_Ano", concat(month("Data"), lit("-"), year("Data")))
cdiReg = cdiReg.select('CDI', 'Mes_Ano')
cdiReg.show(5)

+----+-------+
| CDI|Mes_Ano|
+----+-------+
|6.95| 1-2013|
|6.98| 2-2013|
|7.23| 4-2013|
|7.72| 5-2013|
|8.22| 7-2013|
+----+-------+
only showing top 5 rows



In [0]:
# Extraindo mês e ano da coluna de data
selicReg = selicConcat.withColumn("Mes_Ano", concat(month("Data"), lit("-"), year("Data")))
selicReg = selicReg.select('SELIC', 'Mes_Ano')
selicReg.show(5)

+-----+-------+
|SELIC|Mes_Ano|
+-----+-------+
| 10.4| 1-2014|
|10.65| 2-2014|
|10.65| 3-2014|
| 10.9| 4-2014|
| 10.9| 6-2014|
+-----+-------+
only showing top 5 rows



In [0]:
selicReg.printSchema()

root
 |-- SELIC: double (nullable = true)
 |-- Mes_Ano: string (nullable = true)



In [0]:
igpmReg = igpmFinal.withColumn("Mes_Ano", concat(month("Data"), lit("-"), year("Data")))
igpmReg = igpmReg.select('IGP-M', 'Mes_Ano')
igpmReg.show(5)

+-------+-------+
|  IGP-M|Mes_Ano|
+-------+-------+
|511.977| 1-2013|
|513.467| 2-2013|
|514.526| 3-2013|
|515.276| 4-2013|
|515.299| 5-2013|
+-------+-------+
only showing top 5 rows



In [0]:
igpmReg.printSchema()

root
 |-- IGP-M: float (nullable = true)
 |-- Mes_Ano: string (nullable = true)



In [0]:
inccReg = inccFinal.withColumn('Mes_Ano', concat(month('Data'), lit('-'), year('Data')))
inccReg = inccReg.select('INCC', 'Mes_Ano')
inccReg.show(5)

+-------+-------+
|   INCC|Mes_Ano|
+-------+-------+
|149.224| 7-1996|
|150.334| 8-1996|
|150.564| 9-1996|
|150.915|10-1996|
|151.345|11-1996|
+-------+-------+
only showing top 5 rows



In [0]:
inccReg.printSchema()

root
 |-- INCC: float (nullable = true)
 |-- Mes_Ano: string (nullable = true)



#### Dataset dos Indices

In [0]:
indices = ifixReg.join(cdiReg, "Mes_Ano", "inner").join(selicReg, "Mes_Ano", "inner").join(igpmReg, 'Mes_Ano', 'inner').join(inccReg, 'Mes_Ano', 'inner')
indices.show(5)

+-------+-------+-----+-----+-------+-------+
|Mes_Ano|   IFIX|  CDI|SELIC|  IGP-M|   INCC|
+-------+-------+-----+-----+-------+-------+
| 1-2014|1282.04|10.27| 10.4|540.959|569.745|
| 2-2014|1330.42|10.59|10.65|543.038|572.254|
| 3-2014|1336.07|10.55|10.65|552.087|573.505|
| 4-2014|1350.76| 10.8| 10.9| 556.42|577.322|
| 6-2014|1387.63| 10.8| 10.9|551.554| 592.51|
+-------+-------+-----+-----+-------+-------+
only showing top 5 rows



In [0]:
indices.printSchema()

root
 |-- Mes_Ano: string (nullable = true)
 |-- IFIX: double (nullable = true)
 |-- CDI: double (nullable = true)
 |-- SELIC: double (nullable = true)
 |-- IGP-M: float (nullable = true)
 |-- INCC: float (nullable = true)



In [0]:
indices_date = indices.withColumn("Data", to_date("Mes_Ano", "M-yyyy"))
indices_date.printSchema()


root
 |-- Mes_Ano: string (nullable = true)
 |-- IFIX: double (nullable = true)
 |-- CDI: double (nullable = true)
 |-- SELIC: double (nullable = true)
 |-- IGP-M: float (nullable = true)
 |-- INCC: float (nullable = true)
 |-- Data: date (nullable = true)



In [0]:
indicesFinal = indices_date.select('Data', 'IFIX', 'CDI', 'SELIC', 'INCC', 'IGP-M')
indicesFinal.show(5)

+----------+-------+-----+-----+-------+-------+
|      Data|   IFIX|  CDI|SELIC|   INCC|  IGP-M|
+----------+-------+-----+-----+-------+-------+
|2014-01-01|1282.04|10.27| 10.4|569.745|540.959|
|2014-02-01|1330.42|10.59|10.65|572.254|543.038|
|2014-03-01|1336.07|10.55|10.65|573.505|552.087|
|2014-04-01|1350.76| 10.8| 10.9|577.322| 556.42|
|2014-06-01|1387.63| 10.8| 10.9| 592.51|551.554|
+----------+-------+-----+-----+-------+-------+
only showing top 5 rows



In [0]:
indicesFinal.printSchema()

root
 |-- Data: date (nullable = true)
 |-- IFIX: double (nullable = true)
 |-- CDI: double (nullable = true)
 |-- SELIC: double (nullable = true)
 |-- INCC: float (nullable = true)
 |-- IGP-M: float (nullable = true)



In [0]:
indicesFinal.count()

Out[131]: 85

In [0]:
display(indicesFinal)

Data IFIX CDI SELIC INCC IGP-M 2014-01-01 1282.04 10.27 10.4 569.745 540.959 2014-02-01 1330.42 10.59 10.65 572.254 543.038 2014-03-01 1336.07 10.55 10.65 573.505 552.087 2014-04-01 1350.76 10.8 10.9 577.322 556.42 2014-06-01 1387.63 10.8 10.9 592.51 551.554 2014-07-01 1409.42 10.8 10.9 597.251 548.202 2014-09-01 1421.44 10.81 10.9 599.333 547.839 2014-10-01 1402.46 11.06 11.15 600.549 549.396 2014-12-01 1336.72 11.57 11.65 603.895 558.213 2015-03-01 1347.63 12.6 12.65 613.374 569.536 2015-04-01 1397.06 13.11 13.15 617.36 576.175 2015-06-01 1460.46 13.64 13.65 631.747 582.401 2015-07-01 1470.73 14.13 14.15 635.935 586.426 2015-08-01 1458.12 14.13 14.15 641.019 588.042 2015-09-01 1400.42 14.13 14.15 642.407 593.606 2015-11-01 1452.66 14.14 14.15 646.721 614.051 2015-12-01 1409.01 14.14 14.15 647.492 617.044 2016-02-01 1361.05 14.13 14.15 652.975 632.114 2016-03-01 1485.75 14.13 14.15 658.149 635.349 2016-05-01 1612.36 14.13 14.15 662.089 642.651 2016-06-01 1638.16 14.13 14.15 672.156 653.496 2016-08-01 1767.11 14.13 14.15 681.23 655.602 2016-09-01 1816.05 14.13 14.15 683.764 656.894 2016-10-01 1885.92 13.88 13.9 684.955 657.927 2016-11-01 1837.13 13.88 13.9 686.118 657.752 2017-01-01 1934.69 12.88 12.9 690.614 665.542 2017-03-01 2032.81 12.13 12.15 696.782 666.197 2017-05-01 2056.75 11.13 11.15 697.117 652.758 2017-06-01 2074.14 10.14 10.15 706.596 648.409 2017-07-01 2066.19 9.14 9.15 708.135 643.766 2017-08-01 2084.0 9.14 9.15 710.954 644.383 2017-10-01 2226.25 7.39 7.4 713.354 648.672 2017-11-01 2213.08 7.39 7.4 715.318 652.073 2018-01-01 2285.23 6.89 6.9 718.303 662.826 2018-02-01 2311.41 6.64 6.65 719.332 663.311 2018-04-01 2337.25 6.39 6.4 722.983 671.327 2018-07-01 2154.45 6.39 6.4 735.987 696.8 2018-08-01 2139.4 6.39 6.4 738.158 701.677 2018-10-01 2242.45 6.4 6.4 741.859 718.684 2018-11-01 2300.48 6.4 6.4 743.755 715.166 2018-12-01 2351.59 6.4 6.4 744.699 707.441 2019-01-01 2409.69 6.4 6.4 747.656 707.488 2019-02-01 2434.56 6.4 6.4 749.111 713.747 2019-04-01 2508.45 6.4 6.4 754.203 729.346 2019-05-01 2552.69 6.4 6.4 754.859 732.595 2019-07-01 2659.55 6.4 6.4 765.053 741.346 2019-09-01 2684.26 5.4 5.4 772.31 736.362 2019-10-01 2791.98 4.9 4.9 773.273 741.333 2019-12-01 3197.58 4.4 4.4 775.49 759.112 2020-01-01 3077.28 4.4 4.4 777.47 762.733 2020-03-01 2494.1 3.65 3.65 783.15 771.908 2020-04-01 2603.62 3.65 3.65 784.523 778.101 2020-06-01 2806.28 2.15 2.15 788.616 792.429 2020-07-01 2733.12 2.15 2.15 795.235 810.083 2020-08-01 2782.1 1.9 1.9 801.777 832.313 2020-09-01 2794.88 1.9 1.9 810.965 868.442 2020-11-01 2808.56 1.9 1.9 835.305 925.887 2020-12-01 2870.15 1.9 1.9 842.683 934.758 2021-03-01 2846.77 2.65 2.65 876.75 1011.948 2021-04-01 2861.15 2.65 2.65 885.093 1027.211 2021-05-01 2816.46 3.4 3.4 901.032 1069.289 2021-06-01 2754.89 4.15 4.15 921.762 1075.733 2021-08-01 2749.84 5.15 5.15 938.475 1091.29 2021-09-01 2715.67 6.15 6.15 943.736 1084.312 2021-11-01 2578.4 7.65 7.65 957.984 1091.483 2021-12-01 2804.79 9.15 9.15 960.894 1100.988 2022-01-01 2776.92 9.15 9.15 967.003 1120.999 2022-03-01 2779.97 11.65 11.65 978.717 1161.418 2022-05-01 2820.45 12.65 12.65 1001.923 1183.953 2022-06-01 2795.62 13.15 13.15 1030.105 1190.882 2022-08-01 2976.23 13.65 13.65 1045.469 1185.004 2022-09-01 2990.8 13.65 13.65 1046.507 1173.793 2022-10-01 2991.45 13.65 13.65 1046.956 1162.391 2022-11-01 2867.21 13.65 13.65 1048.429 1155.829 2023-01-01 2821.14 13.65 13.65 1054.561 1163.465 2023-02-01 2808.39 13.65 13.65 1056.775 1162.761 2023-03-01 2760.96 13.65 13.65 1058.65 1163.359 2023-05-01 3013.46 13.65 13.65 1065.289 1131.058 2023-06-01 3155.32 13.65 13.65 1074.292 1109.23 2023-07-01 3197.18 13.65 13.65 1074.895 1101.204 2023-08-01 3212.81 13.15 13.15 1077.497 1099.71 2023-10-01 3155.7 12.65 12.65 1082.251 1109.236 2023-11-01 3176.56 12.15 12.15 1083.335 1115.815 2024-01-01 3333.51 11.65 11.65 1088.663 1124.879 2024-02-01 3360.0 11.15 11.15 1090.871 1119.061

In [0]:
indicesFinal.printSchema()

root
 |-- Data: date (nullable = true)
 |-- IFIX: double (nullable = true)
 |-- CDI: double (nullable = true)
 |-- SELIC: double (nullable = true)
 |-- INCC: float (nullable = true)
 |-- IGP-M: float (nullable = true)



# Iniciando o Processo de Regressão

In [0]:
indicesFinal.describe().show()

+-------+------------------+-----------------+-----------------+-----------------+-----------------+
|summary|              IFIX|              CDI|            SELIC|             INCC|            IGP-M|
+-------+------------------+-----------------+-----------------+-----------------+-----------------+
|  count|                85|               85|               85|               85|               85|
|   mean|2319.9676470588242|9.484117647058808|9.499999999999984| 799.028160903033|814.7069407743567|
| stddev| 630.6641043992679| 4.09802893797379|4.105418892843248|163.1581881050782|225.7878608758626|
|    min|           1282.04|              1.9|              1.9|          569.745|          540.959|
|    max|            3360.0|            14.14|            14.15|         1090.871|         1190.882|
+-------+------------------+-----------------+-----------------+-----------------+-----------------+



#### Assembler incial contendo apenas as variáveis CDI e SELIC

In [0]:
#transformando os dados (linhas) em vetores
from pyspark.ml.feature import VectorAssembler  #importando a biblioteca responsável por criar o vetor a partir da coluna

assembler = VectorAssembler(inputCols=['CDI', 'SELIC'], outputCol='features')  #define o objeto para transformação
indicesRegression = assembler.transform(indicesFinal) #aplica a transformação
indicesRegression.printSchema()

root
 |-- Data: date (nullable = true)
 |-- IFIX: double (nullable = true)
 |-- CDI: double (nullable = true)
 |-- SELIC: double (nullable = true)
 |-- INCC: float (nullable = true)
 |-- IGP-M: float (nullable = true)
 |-- features: vector (nullable = true)



In [0]:
indicesRegression = indicesRegression.select(['features','IFIX'])
display(indicesRegression)

features IFIX Map(vectorType -> dense, length -> 2, values -> List(10.27, 10.4)) 1282.04 Map(vectorType -> dense, length -> 2, values -> List(10.59, 10.65)) 1330.42 Map(vectorType -> dense, length -> 2, values -> List(10.55, 10.65)) 1336.07 Map(vectorType -> dense, length -> 2, values -> List(10.8, 10.9)) 1350.76 Map(vectorType -> dense, length -> 2, values -> List(10.8, 10.9)) 1387.63 Map(vectorType -> dense, length -> 2, values -> List(10.8, 10.9)) 1409.42 Map(vectorType -> dense, length -> 2, values -> List(10.81, 10.9)) 1421.44 Map(vectorType -> dense, length -> 2, values -> List(11.06, 11.15)) 1402.46 Map(vectorType -> dense, length -> 2, values -> List(11.57, 11.65)) 1336.72 Map(vectorType -> dense, length -> 2, values -> List(12.6, 12.65)) 1347.63 Map(vectorType -> dense, length -> 2, values -> List(13.11, 13.15)) 1397.06 Map(vectorType -> dense, length -> 2, values -> List(13.64, 13.65)) 1460.46 Map(vectorType -> dense, length -> 2, values -> List(14.13, 14.15)) 1470.73 Map(vectorType -> dense, length -> 2, values -> List(14.13, 14.15)) 1458.12 Map(vectorType -> dense, length -> 2, values -> List(14.13, 14.15)) 1400.42 Map(vectorType -> dense, length -> 2, values -> List(14.14, 14.15)) 1452.66 Map(vectorType -> dense, length -> 2, values -> List(14.14, 14.15)) 1409.01 Map(vectorType -> dense, length -> 2, values -> List(14.13, 14.15)) 1361.05 Map(vectorType -> dense, length -> 2, values -> List(14.13, 14.15)) 1485.75 Map(vectorType -> dense, length -> 2, values -> List(14.13, 14.15)) 1612.36 Map(vectorType -> dense, length -> 2, values -> List(14.13, 14.15)) 1638.16 Map(vectorType -> dense, length -> 2, values -> List(14.13, 14.15)) 1767.11 Map(vectorType -> dense, length -> 2, values -> List(14.13, 14.15)) 1816.05 Map(vectorType -> dense, length -> 2, values -> List(13.88, 13.9)) 1885.92 Map(vectorType -> dense, length -> 2, values -> List(13.88, 13.9)) 1837.13 Map(vectorType -> dense, length -> 2, values -> List(12.88, 12.9)) 1934.69 Map(vectorType -> dense, length -> 2, values -> List(12.13, 12.15)) 2032.81 Map(vectorType -> dense, length -> 2, values -> List(11.13, 11.15)) 2056.75 Map(vectorType -> dense, length -> 2, values -> List(10.14, 10.15)) 2074.14 Map(vectorType -> dense, length -> 2, values -> List(9.14, 9.15)) 2066.19 Map(vectorType -> dense, length -> 2, values -> List(9.14, 9.15)) 2084.0 Map(vectorType -> dense, length -> 2, values -> List(7.39, 7.4)) 2226.25 Map(vectorType -> dense, length -> 2, values -> List(7.39, 7.4)) 2213.08 Map(vectorType -> dense, length -> 2, values -> List(6.89, 6.9)) 2285.23 Map(vectorType -> dense, length -> 2, values -> List(6.64, 6.65)) 2311.41 Map(vectorType -> dense, length -> 2, values -> List(6.39, 6.4)) 2337.25 Map(vectorType -> dense, length -> 2, values -> List(6.39, 6.4)) 2154.45 Map(vectorType -> dense, length -> 2, values -> List(6.39, 6.4)) 2139.4 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2242.45 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2300.48 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2351.59 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2409.69 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2434.56 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2508.45 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2552.69 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2659.55 Map(vectorType -> dense, length -> 2, values -> List(5.4, 5.4)) 2684.26 Map(vectorType -> dense, length -> 2, values -> List(4.9, 4.9)) 2791.98 Map(vectorType -> dense, length -> 2, values -> List(4.4, 4.4)) 3197.58 Map(vectorType -> dense, length -> 2, values -> List(4.4, 4.4)) 3077.28 Map(vectorType -> dense, length -> 2, values -> List(3.65, 3.65)) 2494.1 Map(vectorType -> dense, length -> 2, values -> List(3.65, 3.65)) 2603.62 Map(vectorType -> dense, length -> 2, values -> List(2.15, 2.15)) 2806.28 Map(vectorType -> dense, length -> 2, values -> Lis

In [0]:
train, test = indicesRegression.randomSplit([0.7, 0.3], seed=123)

Criando o Modelo de Regressão

In [0]:
train.show(5)

+-----------+-------+
|   features|   IFIX|
+-----------+-------+
|  [1.9,1.9]| 2782.1|
|  [1.9,1.9]|2794.88|
|  [1.9,1.9]|2870.15|
|[2.15,2.15]|2733.12|
|[2.15,2.15]|2806.28|
+-----------+-------+
only showing top 5 rows



In [0]:
from pyspark.ml.regression import LinearRegression  #biblioteca que contém o modelo de regressão

lr = LinearRegression(maxIter=10, labelCol='IFIX') # define o objeto a ser utilizado para regressão
lrModel = lr.fit(train)

In [0]:
#Coeficientes angulares e lineares (a e b) da reta de regressão
print(f'Intercepto: {lrModel.intercept}\nCoeficiente Angular: {lrModel.coefficients.values}')

Intercepto: 2812.8810682940625
Coeficiente Angular: [ 14296.34076984 -14320.67830101]


In [0]:
#print das estatísticas do modelo
modelsummary = lrModel.summary

print("Variância Explicada:", modelsummary.explainedVariance)
print('R_2: ', modelsummary.r2)
print('Erro médio quadrático: ',modelsummary.meanSquaredError)

modelsummary.residuals.show(5)


Variância Explicada: 227253.48827756057
R_2:  0.5567481728946275
Erro médio quadrático:  180926.54592828883
+------------------+
|         residuals|
+------------------+
|15.460240941359189|
| 28.24024094135939|
|103.51024094135937|
|-27.43537626450916|
| 45.72462373549115|
+------------------+
only showing top 5 rows



Realizando a Previsão Através do Modelo

In [0]:
modelsummary.predictions.show(5)

+-----------+-------+------------------+
|   features|   IFIX|        prediction|
+-----------+-------+------------------+
|  [1.9,1.9]| 2782.1|2766.6397590586407|
|  [1.9,1.9]|2794.88|2766.6397590586407|
|  [1.9,1.9]|2870.15|2766.6397590586407|
|[2.15,2.15]|2733.12| 2760.555376264509|
|[2.15,2.15]|2806.28| 2760.555376264509|
+-----------+-------+------------------+
only showing top 5 rows



In [0]:
display(modelsummary.predictions)

features IFIX prediction Map(vectorType -> dense, length -> 2, values -> List(1.9, 1.9)) 2782.1 2766.6397590586407 Map(vectorType -> dense, length -> 2, values -> List(1.9, 1.9)) 2794.88 2766.6397590586407 Map(vectorType -> dense, length -> 2, values -> List(1.9, 1.9)) 2870.15 2766.6397590586407 Map(vectorType -> dense, length -> 2, values -> List(2.15, 2.15)) 2733.12 2760.555376264509 Map(vectorType -> dense, length -> 2, values -> List(2.15, 2.15)) 2806.28 2760.555376264509 Map(vectorType -> dense, length -> 2, values -> List(2.65, 2.65)) 2861.15 2748.3866106762384 Map(vectorType -> dense, length -> 2, values -> List(3.4, 3.4)) 2816.46 2730.1334622938325 Map(vectorType -> dense, length -> 2, values -> List(3.65, 3.65)) 2603.62 2724.0490794997045 Map(vectorType -> dense, length -> 2, values -> List(4.15, 4.15)) 2754.89 2711.880313911434 Map(vectorType -> dense, length -> 2, values -> List(4.9, 4.9)) 2791.98 2693.627165529028 Map(vectorType -> dense, length -> 2, values -> List(5.15, 5.15)) 2749.84 2687.5427827349 Map(vectorType -> dense, length -> 2, values -> List(5.4, 5.4)) 2684.26 2681.4583999407573 Map(vectorType -> dense, length -> 2, values -> List(6.15, 6.15)) 2715.67 2663.2052515583587 Map(vectorType -> dense, length -> 2, values -> List(6.39, 6.4)) 2154.45 2514.1574610658377 Map(vectorType -> dense, length -> 2, values -> List(6.39, 6.4)) 2337.25 2514.1574610658377 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2242.45 2657.1208687642306 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2300.48 2657.1208687642306 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2409.69 2657.1208687642306 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2434.56 2657.1208687642306 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2508.45 2657.1208687642306 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2552.69 2657.1208687642306 Map(vectorType -> dense, length -> 2, values -> List(6.4, 6.4)) 2659.55 2657.1208687642306 Map(vectorType -> dense, length -> 2, values -> List(7.39, 7.4)) 2213.08 2489.8199298892964 Map(vectorType -> dense, length -> 2, values -> List(7.39, 7.4)) 2226.25 2489.8199298892964 Map(vectorType -> dense, length -> 2, values -> List(9.15, 9.15)) 2776.92 2590.192658028735 Map(vectorType -> dense, length -> 2, values -> List(9.15, 9.15)) 2804.79 2590.192658028735 Map(vectorType -> dense, length -> 2, values -> List(10.27, 10.4)) 1282.04 701.2464439792188 Map(vectorType -> dense, length -> 2, values -> List(10.59, 10.65)) 1330.42 1695.9059150736816 Map(vectorType -> dense, length -> 2, values -> List(10.8, 10.9)) 1350.76 1117.967901486098 Map(vectorType -> dense, length -> 2, values -> List(10.8, 10.9)) 1387.63 1117.967901486098 Map(vectorType -> dense, length -> 2, values -> List(10.8, 10.9)) 1409.42 1117.967901486098 Map(vectorType -> dense, length -> 2, values -> List(10.81, 10.9)) 1421.44 1260.9313091844474 Map(vectorType -> dense, length -> 2, values -> List(11.15, 11.15)) 3360.0 2541.517595675667 Map(vectorType -> dense, length -> 2, values -> List(11.57, 11.65)) 1336.72 1385.6415685004126 Map(vectorType -> dense, length -> 2, values -> List(11.65, 11.65)) 2779.97 2529.348830087382 Map(vectorType -> dense, length -> 2, values -> List(11.65, 11.65)) 3333.51 2529.348830087382 Map(vectorType -> dense, length -> 2, values -> List(12.13, 12.15)) 2032.81 2231.253249102398 Map(vectorType -> dense, length -> 2, values -> List(12.15, 12.15)) 3176.56 2517.180064499126 Map(vectorType -> dense, length -> 2, values -> List(12.6, 12.65)) 1347.63 1790.1942604189776 Map(vectorType -> dense, length -> 2, values -> List(12.65, 12.65)) 3155.7 2505.0112989108407 Map(vectorType -> dense, length -> 2, values -> List(12.88, 12.9)) 1934.69 2213.000100719985 Map(vectorType -> dense, length -> 2, values -> List(13.15, 13.15)) 2795.62 2492.8425333225846 Map(vectorType -> dense, length -> 2, values -> List(13.65, 13.65)) 2760.96 2480.6737677343285 Map(

In [0]:
# Grafico de linhas do IFIX + predição para comparação
predictionsDF = modelsummary.predictions
predictionsPlot = predictionsDF.join(ifixReg, "IFIX", "inner")
predictionsPlot = predictionsPlot.select('Mes_Ano', 'IFIX', 'prediction')
display(predictionsPlot)

Mes_Ano IFIX prediction 8-2020 2782.1 2766.6397590586407 9-2020 2794.88 2766.6397590586407 12-2020 2870.15 2766.6397590586407 7-2020 2733.12 2760.555376264509 6-2020 2806.28 2760.555376264509 4-2021 2861.15 2748.3866106762384 5-2021 2816.46 2730.1334622938325 4-2020 2603.62 2724.0490794997045 6-2021 2754.89 2711.880313911434 10-2019 2791.98 2693.627165529028 8-2021 2749.84 2687.5427827349 9-2019 2684.26 2681.4583999407573 9-2021 2715.67 2663.2052515583587 7-2018 2154.45 2514.1574610658377 4-2018 2337.25 2514.1574610658377 10-2018 2242.45 2657.1208687642306 11-2018 2300.48 2657.1208687642306 1-2019 2409.69 2657.1208687642306 2-2019 2434.56 2657.1208687642306 4-2019 2508.45 2657.1208687642306 5-2019 2552.69 2657.1208687642306 7-2019 2659.55 2657.1208687642306 11-2017 2213.08 2489.8199298892964 10-2017 2226.25 2489.8199298892964 1-2022 2776.92 2590.192658028735 12-2021 2804.79 2590.192658028735 1-2014 1282.04 701.2464439792188 2-2014 1330.42 1695.9059150736816 4-2014 1350.76 1117.967901486098 6-2014 1387.63 1117.967901486098 7-2014 1409.42 1117.967901486098 9-2014 1421.44 1260.9313091844474 2-2024 3360.0 2541.517595675667 12-2014 1336.72 1385.6415685004126 3-2022 2779.97 2529.348830087382 1-2024 3333.51 2529.348830087382 3-2017 2032.81 2231.253249102398 11-2023 3176.56 2517.180064499126 3-2015 1347.63 1790.1942604189776 10-2023 3155.7 2505.0112989108407 1-2017 1934.69 2213.000100719985 6-2022 2795.62 2492.8425333225846 3-2023 2760.96 2480.6737677343285 2-2023 2808.39 2480.6737677343285 11-2022 2867.21 2480.6737677343285 8-2022 2976.23 2480.6737677343285 10-2022 2991.45 2480.6737677343285 5-2023 3013.46 2480.6737677343285 6-2023 3155.32 2480.6737677343285 7-2023 3197.18 2480.6737677343285 11-2016 1837.13 2188.6625695434436 2-2016 1361.05 2182.5781867493156 9-2015 1400.42 2182.5781867493156 8-2015 1458.12 2182.5781867493156 3-2016 1485.75 2182.5781867493156 8-2016 1767.11 2182.5781867493156 9-2016 1816.05 2182.5781867493156 12-2015 1409.01 2325.541594447694 11-2015 1452.66 2325.541594447694

Databricks visualization. Run in Databricks to view.

#### Assembler adicionando as variáveis IGP-M e INCC

In [0]:
from pyspark.ml.feature import VectorAssembler  #importando a biblioteca responsável por criar o vetor a partir da coluna

assembler2 = VectorAssembler(inputCols=['CDI', 'SELIC' , 'IGP-M', 'INCC'], outputCol='features')  #define o objeto para transformação
indicesRegression = assembler2.transform(indicesFinal) #aplica a transformação
indicesRegression.printSchema()

root
 |-- Data: date (nullable = true)
 |-- IFIX: double (nullable = true)
 |-- CDI: double (nullable = true)
 |-- SELIC: double (nullable = true)
 |-- INCC: float (nullable = true)
 |-- IGP-M: float (nullable = true)
 |-- features: vector (nullable = true)



In [0]:
indicesRegression = indicesRegression.select(['features','IFIX'])
display(indicesRegression)

features IFIX Map(vectorType -> dense, length -> 4, values -> List(10.27, 10.4, 540.958984375, 569.7449951171875)) 1282.04 Map(vectorType -> dense, length -> 4, values -> List(10.59, 10.65, 543.0380249023438, 572.2540283203125)) 1330.42 Map(vectorType -> dense, length -> 4, values -> List(10.55, 10.65, 552.0869750976562, 573.5050048828125)) 1336.07 Map(vectorType -> dense, length -> 4, values -> List(10.8, 10.9, 556.4199829101562, 577.322021484375)) 1350.76 Map(vectorType -> dense, length -> 4, values -> List(10.8, 10.9, 551.5540161132812, 592.510009765625)) 1387.63 Map(vectorType -> dense, length -> 4, values -> List(10.8, 10.9, 548.2020263671875, 597.2509765625)) 1409.42 Map(vectorType -> dense, length -> 4, values -> List(10.81, 10.9, 547.8389892578125, 599.3330078125)) 1421.44 Map(vectorType -> dense, length -> 4, values -> List(11.06, 11.15, 549.39599609375, 600.5490112304688)) 1402.46 Map(vectorType -> dense, length -> 4, values -> List(11.57, 11.65, 558.2130126953125, 603.89501953125)) 1336.72 Map(vectorType -> dense, length -> 4, values -> List(12.6, 12.65, 569.5360107421875, 613.3740234375)) 1347.63 Map(vectorType -> dense, length -> 4, values -> List(13.11, 13.15, 576.1749877929688, 617.3599853515625)) 1397.06 Map(vectorType -> dense, length -> 4, values -> List(13.64, 13.65, 582.4010009765625, 631.7470092773438)) 1460.46 Map(vectorType -> dense, length -> 4, values -> List(14.13, 14.15, 586.426025390625, 635.9349975585938)) 1470.73 Map(vectorType -> dense, length -> 4, values -> List(14.13, 14.15, 588.0419921875, 641.0189819335938)) 1458.12 Map(vectorType -> dense, length -> 4, values -> List(14.13, 14.15, 593.6060180664062, 642.406982421875)) 1400.42 Map(vectorType -> dense, length -> 4, values -> List(14.14, 14.15, 614.051025390625, 646.7210083007812)) 1452.66 Map(vectorType -> dense, length -> 4, values -> List(14.14, 14.15, 617.0440063476562, 647.4920043945312)) 1409.01 Map(vectorType -> dense, length -> 4, values -> List(14.13, 14.15, 632.114013671875, 652.9749755859375)) 1361.05 Map(vectorType -> dense, length -> 4, values -> List(14.13, 14.15, 635.3489990234375, 658.1489868164062)) 1485.75 Map(vectorType -> dense, length -> 4, values -> List(14.13, 14.15, 642.6510009765625, 662.0889892578125)) 1612.36 Map(vectorType -> dense, length -> 4, values -> List(14.13, 14.15, 653.4959716796875, 672.156005859375)) 1638.16 Map(vectorType -> dense, length -> 4, values -> List(14.13, 14.15, 655.6019897460938, 681.22998046875)) 1767.11 Map(vectorType -> dense, length -> 4, values -> List(14.13, 14.15, 656.8939819335938, 683.7639770507812)) 1816.05 Map(vectorType -> dense, length -> 4, values -> List(13.88, 13.9, 657.927001953125, 684.9550170898438)) 1885.92 Map(vectorType -> dense, length -> 4, values -> List(13.88, 13.9, 657.7520141601562, 686.1179809570312)) 1837.13 Map(vectorType -> dense, length -> 4, values -> List(12.88, 12.9, 665.5419921875, 690.614013671875)) 1934.69 Map(vectorType -> dense, length -> 4, values -> List(12.13, 12.15, 666.197021484375, 696.781982421875)) 2032.81 Map(vectorType -> dense, length -> 4, values -> List(11.13, 11.15, 652.7579956054688, 697.1170043945312)) 2056.75 Map(vectorType -> dense, length -> 4, values -> List(10.14, 10.15, 648.4089965820312, 706.5960083007812)) 2074.14 Map(vectorType -> dense, length -> 4, values -> List(9.14, 9.15, 643.7659912109375, 708.135009765625)) 2066.19 Map(vectorType -> dense, length -> 4, values -> List(9.14, 9.15, 644.3829956054688, 710.9539794921875)) 2084.0 Map(vectorType -> dense, length -> 4, values -> List(7.39, 7.4, 648.6719970703125, 713.35400390625)) 2226.25 Map(vectorType -> dense, length -> 4, values -> List(7.39, 7.4, 652.072998046875, 715.3179931640625)) 2213.08 Map(vectorType -> dense, length -> 4, values -> List(6.89, 6.9, 662.8259887695312, 718.302978515625)) 2285.23 Map(vectorType -> dense, length -> 4, values -> List(6.64, 6.65, 663.3109741210938, 719.3319702148438)) 2311.41 Map(vectorType -> dense, length -> 4, values -> List(6.39, 6.4, 6

In [0]:
train, test = indicesRegression.randomSplit([0.7, 0.3], seed=123)

In [0]:
from pyspark.ml.feature import VectorAssembler  #importando a biblioteca responsável por criar o vetor a partir da coluna

assembler = VectorAssembler(inputCols=['CDI', 'SELIC'], outputCol='features')  #define o objeto para transformação
indicesRegression = assembler.transform(indicesFinal) #aplica a transformação
indicesRegression.printSchema()

root
 |-- Data: date (nullable = true)
 |-- IFIX: double (nullable = true)
 |-- CDI: double (nullable = true)
 |-- SELIC: double (nullable = true)
 |-- INCC: float (nullable = true)
 |-- IGP-M: float (nullable = true)
 |-- features: vector (nullable = true)



In [0]:
train.show(5)

+--------------------+-------+
|            features|   IFIX|
+--------------------+-------+
|[1.9,1.9,832.3129...| 2782.1|
|[1.9,1.9,868.4420...|2794.88|
|[1.9,1.9,934.7579...|2870.15|
|[2.15,2.15,792.42...|2806.28|
|[2.15,2.15,810.08...|2733.12|
+--------------------+-------+
only showing top 5 rows



In [0]:
from pyspark.ml.regression import LinearRegression  #biblioteca que contém o modelo de regressão

lr = LinearRegression(maxIter=10, labelCol='IFIX') # define o objeto a ser utilizado para regressão
lrModel = lr.fit(train)

In [0]:
#Coeficientes angulares e lineares (a e b) da reta de regressão
print(f'Intercepto: {lrModel.intercept}\nCoeficiente Angular: {lrModel.coefficients.values}')

Intercepto: -199.3859719802086
Coeficiente Angular: [ 1.82857430e+03 -1.88466487e+03 -1.65588574e+00  5.53768299e+00]


In [0]:
#print das estatísticas do modelo
modelsummary = lrModel.summary

print("Variância Explicada:", modelsummary.explainedVariance)
print('R_2: ', modelsummary.r2)
print('Erro médio quadrático: ',modelsummary.meanSquaredError)

modelsummary.residuals.show(5)


Variância Explicada: 395461.63281228987
R_2:  0.9715550907042869
Erro médio quadrático:  11578.211449745791
+-----------------+
|        residuals|
+-----------------+
|26.28653699805045|
|48.01157548721039|
|57.44927120809734|
|71.32704618709658|
|-9.25364912225723|
+-----------------+
only showing top 5 rows



In [0]:
modelsummary.predictions.show(5)

+--------------------+-------+------------------+
|            features|   IFIX|        prediction|
+--------------------+-------+------------------+
|[1.9,1.9,832.3129...| 2782.1|2755.8134630019495|
|[1.9,1.9,868.4420...|2794.88|2746.8684245127897|
|[1.9,1.9,934.7579...|2870.15|2812.7007287919027|
|[2.15,2.15,792.42...|2806.28|2734.9529538129036|
|[2.15,2.15,810.08...|2733.12| 2742.373649122257|
+--------------------+-------+------------------+
only showing top 5 rows



In [0]:
display(modelsummary.predictions)

features IFIX prediction Map(vectorType -> dense, length -> 4, values -> List(1.9, 1.9, 832.31298828125, 801.7769775390625)) 2782.1 2755.8134630019495 Map(vectorType -> dense, length -> 4, values -> List(1.9, 1.9, 868.4420166015625, 810.9650268554688)) 2794.88 2746.8684245127897 Map(vectorType -> dense, length -> 4, values -> List(1.9, 1.9, 934.7579956054688, 842.6829833984375)) 2870.15 2812.7007287919027 Map(vectorType -> dense, length -> 4, values -> List(2.15, 2.15, 792.4290161132812, 788.6160278320312)) 2806.28 2734.9529538129036 Map(vectorType -> dense, length -> 4, values -> List(2.15, 2.15, 810.0830078125, 795.2349853515625)) 2733.12 2742.373649122257 Map(vectorType -> dense, length -> 4, values -> List(2.65, 2.65, 1027.2110595703125, 885.093017578125)) 2861.15 2852.3944149502286 Map(vectorType -> dense, length -> 4, values -> List(3.4, 3.4, 1069.2889404296875, 901.031982421875)) 2816.46 2828.9152582087345 Map(vectorType -> dense, length -> 4, values -> List(3.65, 3.65, 778.1010131835938, 784.5230102539062)) 2603.62 2651.876799301356 Map(vectorType -> dense, length -> 4, values -> List(4.15, 4.15, 1075.7330322265625, 921.7620239257812)) 2754.89 2890.9730484763695 Map(vectorType -> dense, length -> 4, values -> List(4.9, 4.9, 741.3330078125, 773.2730102539062)) 2791.98 2580.3482678287533 Map(vectorType -> dense, length -> 4, values -> List(5.15, 5.15, 1091.2900390625, 938.4749755859375)) 2749.84 2901.6728798083223 Map(vectorType -> dense, length -> 4, values -> List(5.4, 5.4, 736.3619995117188, 772.3099975585938)) 2684.26 2555.201545091253 Map(vectorType -> dense, length -> 4, values -> List(6.15, 6.15, 1084.31201171875, 943.7360229492188)) 2715.67 2886.27113732147 Map(vectorType -> dense, length -> 4, values -> List(6.39, 6.4, 696.7999877929688, 735.9869995117188)) 2154.45 2345.1901537958297 Map(vectorType -> dense, length -> 4, values -> List(6.39, 6.4, 701.677001953125, 738.1580200195312)) 2139.4 2349.136798917959 Map(vectorType -> dense, length -> 4, values -> List(6.4, 6.4, 707.4409790039062, 744.698974609375)) 2351.59 2394.0997875040684 Map(vectorType -> dense, length -> 4, values -> List(6.4, 6.4, 707.4879760742188, 747.656005859375)) 2409.69 2410.397067381704 Map(vectorType -> dense, length -> 4, values -> List(6.4, 6.4, 715.166015625, 743.7550048828125)) 2300.48 2376.0806043998 Map(vectorType -> dense, length -> 4, values -> List(6.4, 6.4, 718.6840209960938, 741.8590087890625)) 2242.45 2359.755764142766 Map(vectorType -> dense, length -> 4, values -> List(6.4, 6.4, 729.3460083007812, 754.2030029296875)) 2508.45 2410.4578577628345 Map(vectorType -> dense, length -> 4, values -> List(6.4, 6.4, 732.594970703125, 754.8590087890625)) 2552.69 2408.710699730291 Map(vectorType -> dense, length -> 4, values -> List(6.4, 6.4, 741.3460083007812, 765.052978515625)) 2659.55 2450.670954098464 Map(vectorType -> dense, length -> 4, values -> List(7.39, 7.4, 648.6719970703125, 713.35400390625)) 2226.25 2243.4596817043684 Map(vectorType -> dense, length -> 4, values -> List(7.39, 7.4, 652.072998046875, 715.3179931640625)) 2213.08 2248.703962582129 Map(vectorType -> dense, length -> 4, values -> List(9.15, 9.15, 1100.988037109375, 960.8939819335938)) 2804.79 2785.4011693174643 Map(vectorType -> dense, length -> 4, values -> List(9.15, 9.15, 1120.9990234375, 967.0029907226562)) 2776.92 2786.0950164133983 Map(vectorType -> dense, length -> 4, values -> List(10.27, 10.4, 540.958984375, 569.7449951171875)) 1282.04 1238.858329156654 Map(vectorType -> dense, length -> 4, values -> List(10.59, 10.65, 543.0380249023438, 572.2540283203125)) 1330.42 1363.2874645621546 Map(vectorType -> dense, length -> 4, values -> List(10.8, 10.9, 548.2020263671875, 597.2509765625)) 1409.42 1405.9960283878368 Map(vectorType -> dense, length -> 4, values -> List(10.8, 10.9, 551.5540161132812, 592.510009765625)) 1387.63 1374.1915451654033 Map(vectorType -> dense, length -> 4, values -> List(10.8, 10.9, 556.4199829101562, 577.322021484375)) 1350.76 1282.02779

In [0]:
# Grafico de linhas do IFIX + predição para comparação
predictionsDF = modelsummary.predictions
predictionsPlot = predictionsDF.join(ifixReg, "IFIX", "inner")
predictionsPlot = predictionsPlot.select('Mes_Ano', 'IFIX', 'prediction')
display(predictionsPlot)

Mes_Ano IFIX prediction 8-2020 2782.1 2755.8134630019495 9-2020 2794.88 2746.8684245127897 12-2020 2870.15 2812.7007287919027 6-2020 2806.28 2734.9529538129036 7-2020 2733.12 2742.373649122257 4-2021 2861.15 2852.3944149502286 5-2021 2816.46 2828.9152582087345 4-2020 2603.62 2651.876799301356 6-2021 2754.89 2890.9730484763695 10-2019 2791.98 2580.3482678287533 8-2021 2749.84 2901.6728798083223 9-2019 2684.26 2555.201545091253 9-2021 2715.67 2886.27113732147 7-2018 2154.45 2345.1901537958297 8-2018 2139.4 2349.136798917959 12-2018 2351.59 2394.0997875040684 1-2019 2409.69 2410.397067381704 11-2018 2300.48 2376.0806043998 10-2018 2242.45 2359.755764142766 4-2019 2508.45 2410.4578577628345 5-2019 2552.69 2408.710699730291 7-2019 2659.55 2450.670954098464 10-2017 2226.25 2243.4596817043684 11-2017 2213.08 2248.703962582129 12-2021 2804.79 2785.4011693174643 1-2022 2776.92 2786.0950164133983 1-2014 1282.04 1238.858329156654 2-2014 1330.42 1363.2874645621546 7-2014 1409.42 1405.9960283878368 6-2014 1387.63 1374.1915451654033 4-2014 1350.76 1282.027795750876 9-2014 1421.44 1436.412548433167 2-2024 3360.0 3363.064572871634 12-2014 1336.72 1420.7151404224019 1-2024 3333.51 3313.1581081423365 3-2022 2779.97 2643.807792743236 3-2017 2032.81 1837.9536869573512 11-2023 3176.56 3270.6171237500657 3-2015 1347.63 1453.2239262081735 5-2022 2820.45 2678.909216280388 1-2017 1934.69 1762.8141567636385 8-2023 3212.81 3208.865496720563 7-2023 3197.18 3163.937630841392 6-2023 3155.32 3147.308115985865 11-2022 2867.21 2926.9241971704423 10-2022 2991.45 2907.9018000791357 3-2023 2760.96 2971.056370821959 1-2023 2821.14 2948.237391822057 9-2022 2990.8 2886.5344729849166 8-2022 2976.23 2862.222324425202 11-2016 1837.13 1694.7252954495175 7-2015 1470.73 1520.9128870277787 8-2015 1458.12 1546.390524446371 9-2015 1400.42 1544.863440014491 3-2016 1485.75 1562.916062950317 8-2016 1767.11 1657.194650316772 9-2016 1816.05 1669.087728644327 11-2015 1452.66 1553.1842946327888 12-2015 1409.01 1552.497792091037

Databricks visualization. Run in Databricks to view.

# Conclusão

Analisando as variáveis podemos concluir que adicionando as variáveis IGP-M e INCC ao conjunto, acaba melhorando o fit do modelo. Temos então, que o resultado de R² indica que aproximadamente 97% da variabilidade na variável dependente (IFIX) pode ser explicada pelas variáveis independentes (indices) incluídas no modelo.